tests/phpunit/DiffOp/Diff/DiffTest.php

```diff
@@ -673,4 +672,54 @@ public function invalidDiffOpsProvider() {
 		return $this->arrayWrap( $diffOpLists );
 	}
 
+	/**
+	 * @dataProvider equalsProvider
+	 */
+	public function testEquals( Diff $diff, Diff $target ) {
+		$this->assertTrue( $diff->equals( $target ) );
+		$this->assertTrue( $target->equals( $diff ) );
+	}
+
+	public function equalsProvider() {
+		return array(
+			// Empty diffs
+			array( new Diff(), new Diff() ),
+			array( new Diff(), new Diff( array(), null ) ),
+
+			// Simple diffs
+			array( new Diff( array( new DiffOpAdd( 1 ) ) ), new Diff( array( new DiffOpAdd( 1 ) ) ) ),
+			array( new Diff( array( new DiffOpAdd( 1 ) ) ), new Diff( array( new DiffOpAdd( '1' ) ) ) ),
+
+			// Subclasses that are shortcuts and should be equal
+			array( new Diff( array(), false ), new ListDiff() ),
+			array( new Diff( array(), true ), new MapDiff() ),
+		);
+	}
+
+	/**
+	 * @dataProvider notEqualsProvider
+	 */
+	public function testNotEquals( Diff $diff, Diff $target ) {
+		$this->assertFalse( $diff->equals( $target ) );
+		$this->assertFalse( $target->equals( $diff ) );
+	}
+
+	public function notEqualsProvider() {
+		return array(
+			// Empty diffs
+			array( new Diff(), new Diff( array(), false ) ),
+			array( new Diff(), new Diff( array(), true ) ),
+
+			// Simple diffs
+			array( new Diff(), new Diff( array( new DiffOpAdd( 1 ) ) ) ),
+			array( new Diff( array( new DiffOpAdd( 1 ) ) ), new Diff( array( new DiffOpRemove( 1 ) ) ) ),
+			array( new Diff( array( new DiffOpAdd( 1 ) ) ), new Diff( array( new DiffOpAdd( 2 ) ) ) ),
```

Comment:
This single line is the relevant test for [T92487](https://phabricator.wikimedia.org/T92487).